src/network_event/client_event.rs

```diff
@@ -119,155 +119,219 @@ impl ClientEventAppExt for App {
+            for (client_id, message) in server.receive(channel_id) {
+                match DefaultOptions::new().deserialize(&message) {
+                    Ok(event) => {
+                        trace!(
+                            "applying event `{}` from `{client_id:?}`",
+                            std::any::type_name::<T>()
+                        );
+                        client_events.send(FromClient { client_id, event });
+                    }
+                    Err(e) => debug!("unable to deserialize event from {client_id:?}: {e}"),
+                }
+            }
+        })
+    })
 }
 
 /// Discards all pending events.
 ///
 /// We discard events while waiting to connect to ensure clean reconnects.
-fn reset<T: Event>(mut events: ResMut<Events<T>>) {
-    let drained_count = events.drain().count();
-    if drained_count > 0 {
-        warn!("discarded {drained_count} client events due to a disconnect");
+fn reset<T: Event>(world: &mut World) {
+    world.resource_scope(|_world, mut events: Mut<Events<T>>| {
+        let drained_count = events.drain().count();
+        if drained_count > 0 {
+            warn!("Discarded {drained_count} client events due to a disconnect");
+        }
+    })
+}
+
+#[derive(Resource, Default)]
+struct ClientEventRegistry {
+    events: Vec<ClientEventFns>,
+}
+
+impl ClientEventRegistry {
+    fn register_event<T: Event + Serialize + DeserializeOwned>(
+        &mut self,
+        channel_id: u8,
+        send: SendFn,
+        receive: ReceiveFn,
+    ) {
+        self.events
+            .push(ClientEventFns::new::<T>(channel_id, send, receive));
     }
 }
 
-/// Holds a client's channel ID for `T`.
-#[derive(Resource)]
-pub struct ClientEventChannel<T> {
-    id: u8,
-    marker: PhantomData<T>,
+fn send_system(world: &mut World) {
+    world.resource_scope(|world, registry: Mut<ClientEventRegistry>| {
+        for event in &registry.events {
+            (event.send)(world, event.channel_id);
+        }
+    });
 }
 
-impl<T> ClientEventChannel<T> {
-    fn new(id: u8) -> Self {
-        Self {
-            id,
-            marker: PhantomData,
+fn resend_locally_system(world: &mut World) {
+    world.resource_scope(|world, registry: Mut<ClientEventRegistry>| {
+        for event in &registry.events {
+            (event.resend_locally)(world);
         }
-    }
+    });
 }
 
-impl<T> Clone for ClientEventChannel<T> {
-    fn clone(&self) -> Self {
-        *self
-    }
+fn receive_system(world: &mut World) {
+    world.resource_scope(|world, registry: Mut<ClientEventRegistry>| {
+        for event in &registry.events {
+            (event.receive)(world, event.channel_id);
+        }
+    });
+}
+
+fn reset_system(world: &mut World) {
+    world.resource_scope(|world, registry: Mut<ClientEventRegistry>| {
+        for event in &registry.events {
+            (event.reset)(world);
+        }
+    });
 }
 
-impl<T> Copy for ClientEventChannel<T> {}
+pub struct ClientNetWorkEventPlugin;
```

Comment:
Let's keep names shorter:
```suggestion
pub struct ClientEventPlugin;
```
Also could you move systems under `impl ClientEventPlugin` for consistency with the rest of the codebase?